src/main/java/org/opentripplanner/street/search/state/State.java

```diff
@@ -10,9 +10,9 @@
 import org.opentripplanner.ext.dataoverlay.routing.DataOverlayContext;
 import org.opentripplanner.framework.tostring.ToStringBuilder;
 import org.opentripplanner.routing.api.request.preference.RoutingPreferences;
+import org.opentripplanner.service.vehiclerental.street.VehicleRentalEdge;
+import org.opentripplanner.service.vehiclerental.street.VehicleRentalPlaceVertex;
```

Comment:
This creates a circular dependency, but I don't know really what would be the best way to solve it

Reply:
So something like `GraphPathReverser`?

Reply:
Yes, exactly

Reply:
I looked at the `State#reverse()`, to me it seems like this method should be where it is, coping it outside would just make the code more fragmented. Separating data from behaviour to fix cyclic dependencies just makes it worse. The real solution would be to analyse this and define a model without cyclic dependencies and then add the proper extension points (interfaces) to resolve the cyclic dependencies. This is not an easy task. The `orig(this)` is used frequently. There seems to be a circular dependency between Edge <-> Vertex <-> State, but it might be possible that the reverse method should delegate more to the V/E and then the specific E/V could override methods to avoid the dependency from `org.opentripplanner.street` to the `org.opentripplanner.service.vehiclerental.street` package. There is also a dependency from `o.o.model.plan.place` witch is not allowed. 

Overall this is a very nice refactoring that moves the code in the right direction, fixing the last bit here might be a lot of work - to do it properly - so I suggest merging this, leaving the reminding work for another PR.

Reply:
I would prefer to merge it as is as well.

Reply:
We can merge it as is. One idea for the reversing is to create a method `reverseTraverse` on the `Edge` types, and delegate the responsibility to each of the types requiring special attention later on.